include/nbl/ext/Scan/Scan.h

```diff
@@ -0,0 +1,190 @@
+#ifndef _NBL_EXT_SCAN_INCLUDED_
+
+#include "nabla.h"
+
+namespace nbl
+{
+namespace ext
+{
+namespace Scan
+{
+
+typedef uint32_t uint;
+#include "nbl/builtin/glsl/ext/Scan/parameters_struct.glsl"
+
+template <typename T>
+class Scan final : public core::IReferenceCounted
+{
+public:
+	_NBL_STATIC_INLINE_CONSTEXPR uint32_t DEFAULT_WORKGROUP_SIZE = 256u;
+
+	enum Operator : uint32_t
+	{
+		AND = 1 << 0,
+		XOR = 1 << 1,
+		OR  = 1 << 2,
+		ADD = 1 << 3,
+		MUL = 1 << 4,
+		MIN = 1 << 5,
+		MAX = 1 << 6,
+	};
+
+	typedef nbl_glsl_ext_Scan_Parameters_t Parameters_t;
+
+	struct DispatchInfo_t
+	{
+		uint32_t upsweep_pass_count;
+		uint32_t downsweep_pass_count;
+		uint32_t wg_count;
+		std::stack<uint32_t> element_count_pass_stack;
```

Comment:
the only thing you should keep is `uint32_t wg_count[3];`

btw AMD and a few other GPUs have a "weird" limit on number of workgroups in X,Y,Z dimensions, about 64k

You need to prepare the code for the eventuality of having to use a new utility function in `core` that can split a `T` (either `uint32_t` or `uint64_t`) into 3 factors which when multiplied are either greater than or equal to the input. You'll need to use it if 

For example a 64M input would be factored into (64k,1024,1).

P.S. For a later implementor, an approximate 3-factor algorithm would have to do the following:
- produce 3 natural (integer greater than 0) numbers less than or equal to the Maximum
- start at each number equal to `ceil(pow(input_size,1.f/3.f))`
- explore each candidate triple in the surface `w*d*h = input_size` (iterate in 2D over w,h in the range `[1,MAX]`)

Reply:
strange